src/OrleansProviders/Streams/Common/PooledCache/IObjectPool.cs

```diff
@@ -41,12 +66,18 @@ public abstract class PooledResource<T> : IDisposable
         public IObjectPool<T> Pool { set { pool = value; } }
 
         /// <summary>
+        /// Identification
+        /// </summary>
+        public virtual object Id { get; }
+
+        /// <summary>
         /// If this object is to be used in a fixed size object pool, this call should be
         ///   overridden with the purge implementation that returns the object to the pool.
         /// </summary>
-        public virtual void SignalPurge()
+        public virtual void SignalPurge(IDisposable purgeCandidate)
         {
-            Dispose();
+            if(purgeCandidate == this)
```

Comment:
== will do that in this circumstance, but ReferenceEquals is definitely safer. will do